On Apple platforms, use swiftmodule directories for the stdlib

jrose-apple opened the pull request:
This changes the Swift resource directory from looking like

    lib/
      swift/
        macosx/
          libswiftCore.dylib
          libswiftDarwin.dylib
          x86_64/
            Swift.swiftmodule
            Swift.swiftdoc
            Darwin.swiftmodule
            Darwin.swiftdoc

to

    lib/
      swift/
        macosx/
          libswiftCore.dylib
          libswiftDarwin.dylib
          Swift.swiftmodule/
            x86_64.swiftmodule
            x86_64.swiftdoc
          Darwin.swiftmodule/
            x86_64.swiftmodule
            x86_64.swiftdoc

matching the layout we use for multi-architecture swiftmodules everywhere else (particularly frameworks).

There's no change in this commit to how Linux swiftmodules are packaged. There's been past interest in going the *opposite* direction for Linux, since there's not standard support for fat (multi-architecture) .so libraries. Moving the .so search path *down* to an architecture-specific directory on Linux would allow the same resource directory to be used for both host-compiling and cross-compiling.

rdar://problem/43545560

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
apple/swift-lldb#1203
@swift-ci Please clean test

jrose-apple commented:
A little bit further every time…

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
Oops.

apple/swift-lldb#1203
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Grr, right, that was the test I didn't update properly.

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please clean test macOS

jrose-apple commented:
@swift-ci Please clean test macOS

jrose-apple commented:
@swift-ci Please build toolchain

jrose-apple commented:
Cool, all files seem to be ending up where expected in the Apple toolchain, no change right now to the Linux toolchain.

jrose-apple commented:
I haven't actually tested this with Apple's unreleased architectures (like arm64e), so I can't say this is ready to merge yet. But it is ready for review!

(…if non-urgently. This is obviously not a change needed for Swift 5.0.)

jrose-apple commented:
@swift-ci Please test

bob-wilson reviewed (approved):
I don't have much expertise in any of this code, but the overall change looks good. Thank you for doing this!

jrose-apple commented:
@swift-ci Please clean test

jrose-apple committed:
On Apple platforms, use swiftmodule directories for the stdlib

This changes the Swift resource directory from looking like

    lib/
      swift/
        macosx/
          libswiftCore.dylib
          libswiftDarwin.dylib
          x86_64/
            Swift.swiftmodule
            Swift.swiftdoc
            Darwin.swiftmodule
            Darwin.swiftdoc

to

    lib/
      swift/
        macosx/
          libswiftCore.dylib
          libswiftDarwin.dylib
          Swift.swiftmodule/
            x86_64.swiftmodule
            x86_64.swiftdoc
          Darwin.swiftmodule/
            x86_64.swiftmodule
            x86_64.swiftdoc

matching the layout we use for multi-architecture swiftmodules
everywhere else (particularly frameworks).

There's no change in this commit to how Linux swiftmodules are
packaged. There's been past interest in going the /opposite/ direction
for Linux, since there's not standard support for fat
(multi-architecture) .so libraries. Moving the .so search path /down/
to an architecture-specific directory on Linux would allow the same
resource directory to be used for both host-compiling and
cross-compiling.

rdar://problem/43545560

compnerd commented on cmake/modules/SwiftSource.cmake:
```diff
@@ -317,19 +317,32 @@ function(_compile_swift_files
 
     set(module_base "${module_dir}/${SWIFTFILE_MODULE_NAME}")
+    if(SWIFTFILE_SDK IN_LIST SWIFT_APPLE_PLATFORMS)
+      set(specific_module_dir "${module_base}.swiftmodule")
```
You only use this once below, I think you could just inline it there as it is already covered by the conditional.

jrose-apple replied:
It's used to form `module_base` as well as being passed to the `install` command. I didn't want the `install` to have to know about the particular extension here.

compnerd commented on test/SIL/Serialization/deserialize_stdlib.sil:
```diff
@@ -1,5 +1,5 @@
 // Make sure that we can deserialize the stdlib.
-// RUN: %target-sil-opt %platform-module-dir/Swift.swiftmodule > /dev/null
-// RUN: llvm-bcanalyzer %platform-module-dir/Swift.swiftmodule | %FileCheck %s
+// RUN: %target-sil-opt $(if [[ -d %platform-module-dir/Swift.swiftmodule ]]; then echo %platform-module-dir/Swift.swiftmodule/%target-swiftmodule-name; else echo %platform-module-dir/Swift.swiftmodule; fi) -module-name Swift > /dev/null
```
Can we split this up someway please?  This will break with Windows, as the subshell syntax is not portable.

jrose-apple replied:
Hm. It may just need another substitution, then.

compnerd commented on test/lit.cfg:
```diff
@@ -1327,10 +1327,12 @@ if hasattr(config, 'target_swift_autolink_extract'):
+if run_vendor != 'apple':
+    platform_module_dir = make_path(platform_module_dir, run_cpu)
 lit_config.note('Using platform module dir: ' + platform_module_dir)
 if test_sdk_overlay_dir is not None:
```
While in the area, mind cleaning this up?  It should read: `if test_sdk_overlay_dir:`

compnerd commented on validation-test/SIL/parse_stdlib.sil:
```diff
@@ -1,5 +1,5 @@
 // RUN: rm -f %t.*
-// RUN: %target-sil-opt -enable-sil-verify-all=true -sil-disable-ast-dump %platform-module-dir/Swift.swiftmodule -module-name=Swift -o %t.sil
+// RUN: %target-sil-opt -enable-sil-verify-all=true -sil-disable-ast-dump $(if [[ -d %platform-module-dir/Swift.swiftmodule ]]; then echo %platform-module-dir/Swift.swiftmodule/%target-swiftmodule-name; else echo %platform-module-dir/Swift.swiftmodule; fi) -module-name=Swift -o %t.sil
```
Can we split this up please?  This syntax is not supported on Windows, and this will cause a regression on the windows target.

compnerd commented on validation-test/SIL/verify_all_overlays.sil:
```diff
@@ -1,4 +1,4 @@
-// RUN: for x in %platform-sdk-overlay-dir/*.swiftmodule; do [[ $(basename "$x") = Swift.swiftmodule ]] && continue; llvm-bcanalyzer $x | %FileCheck %s; %target-sil-opt -sdk %sdk -enable-sil-verify-all $x > /dev/null; done
+// RUN: for x in %platform-sdk-overlay-dir/*.swiftmodule; do [[ $(basename "$x") = Swift.swiftmodule ]] && continue; llvm-bcanalyzer $(if [[ -d "$x" ]]; then echo "$x/%target-swiftmodule-name"; else echo "$x"; fi)  | %FileCheck %s; %target-sil-opt -sdk %sdk -enable-sil-verify-all $(if [[ -d "$x" ]]; then echo "$x/%target-swiftmodule-name"; else echo "$x"; fi) > /dev/null; done
```
I think that this test would previously work with coreutils from GNUWin32, but the subshell syntax will cause a regression.

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
@swift-ci Please smoke test macOS

jrose-apple commented:
@swift-ci Please clean test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented on test/DebugInfo/variables.swift:
```diff
@@ -41,7 +41,7 @@ print(", \(glob_b)", terminator: "")
 var unused: Int32 = -1
 
-// CHECK-DAG: ![[RT:[0-9]+]] ={{.*}}"{{.*}}Swift.swiftmodule"
+// CHECK-DAG: !{{RT:[0-9]+}} = !DIFile({{filename: ".+[.]swiftmodule", directory: ".*Swift[.]swiftmodule"|filename: "Swift[.]swiftmodule", directory: ".*"}})
```
Oops. I missed some square brackets.

jrose-apple commented:
[Debug source compat suite passed](https://ci.swift.org/job/swift-PR-source-compat-suite-debug/748/)
[Release source compat suite still going](https://ci.swift.org/job/swift-PR-source-compat-suite/2520/)

I need to figure out what was wrong with my fixes for the debug info tests, so building locally. Will try again next week.

jrose-apple commented:
@swift-ci Please clean test

paulmillr commented:
@jrose-apple how would it play with ARMOSX branch?

jrose-apple commented:
I'm not sure what you're referring to, but I wouldn't expect this to be a problem. The "right" condition is "does your platform support fat dylibs" but right now it's approximating this with "is your platform an Apple platform in one of these lists".

jrose-apple commented:
@swift-ci Please clean test Linux

jrose-apple commented:
Here we go. I'll be watching this to make sure it works (and to fix the merge conflicts for Apple-internal folks), but if it's breaking things overnight or tomorrow morning please just revert.
